### PR TITLE
study(color): a Q16 bind-time inverse is accurate enough (#4043)

### PR DESCRIPTION
**Pull request by zackees**
Refs #4043 (P9).

The survey on #4043 calls the fixed-point bind-time derivation **"the one that is genuinely blocked on nothing but effort"**, and gives the reason it is not a small change:

> `EmitterProfile` itself stores `float xy_r[2]`, `lum_r` and so on. It is P2's public type and the whole legacy RGBW stack consumes it. A float-free bind-time derivation needs a fixed-point profile representation, which is a cross-cutting P2/P9 change rather than a rewrite of one 129-line file.

That is a statement about the *profile*. It assumes the narrower question is already answered yes: **is a Q16 3×3 inverse accurate enough to replace the float one at all?** If it is not, converting the profile buys nothing — the derivation would still have to reach float to stay inside A1, and item 2 is not "effort" but "impossible as scoped".

It had not been measured. `ci/color_fixed_inverse_study.py` measures it.

## What is actually being compared

Both paths end in s16.16, so this is not a question about the output format. It is a question about where the arithmetic happens — float32 carries a 24-bit mantissa that follows the magnitude, Q16 with 64-bit intermediates carries a fixed 1/65536 step and far more headroom around one, and neither dominates a priori.

| primaries | worst coefficient error | worst ΔE2000 |
|---|---:|---:|
| sRGB | 0 ULP | 0.0000 |
| BT.2020 | 1 ULP | 0.0057 |
| Display P3 | 1 ULP | 0.0064 |

against A1's 0.5 budget, of which the shipped mapper already spends 0.15.

## Where it gives out is the more interesting half

Conditioning is what decides a fixed-point inverse's accuracy, so green was walked onto the red–blue line until the matrix was nearly singular:

| green collapsed | worst coefficient error | worst ΔE2000 |
|---|---:|---:|
| 90% of the way | 6 ULP | 0.0421 |
| 99% | 575 ULP | 0.0270 |
| 99.9% | **28 487 ULP** | **0.0914** |

The coefficients lose **four orders of magnitude** of precision and the colour error does not follow. Those ULPs sit in directions a near-collinear device can barely produce, so they never become visible error: even at the edge of singularity the two derivations stay an order of magnitude inside the budget.

**So the answer is yes**, and P9 item 2 is effort rather than an open question.

## What this does not say

Stated because it is easy to over-read: it compares the two derivations against **each other**, not against float64 — the shipped path is itself Q16-quantised, and "can this replace that" is the question. And every target is in gamut by construction, since the sweep drives are in `[0, 1]`; an out-of-gamut target goes through the mapper, which is measured separately.

## Verification

- `uv run python -m pytest ci/tests/test_color_fixed_inverse.py -q` — 8 passed, 14 subtests
- `bash lint` — 0 errors
- Mutation-tested: truncating the divide instead of rounding, transposing the cofactors, and never reporting singular each fail. The identity check is what catches the transposition — comparing against the float path cannot, because it would be transposed the same way.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

https://claude.ai/code/session_01MLhWkMfzLjrnLTDMBE6Fj9

## Summary by CodeRabbit

* **Documentation**
  * Added an analysis of fixed-point color-gamut matrix inversion accuracy.
  * Documented comparisons across common color spaces and near-singular conditions using the shipped numerical precision.
  * Recorded color-error findings and implementation considerations.

* **Tests**
  * Added coverage for inversion accuracy, rounding behavior, singular matrices, identity residuals, and invalid inputs.
  * Added color-error budget checks across standard and edge-case color configurations.

### CODE REVIEW

File: ci/color_fixed_inverse_study.py [hunk 1]
```diff
@@ -0,0 +1,315 @@
...
+    """Worst-case disagreement between the two derivations on one device.
+
+    Driven from drives rather than from XYZ: a drive vector is what the device
+    is actually asked for, so re-rendering it gives a target that is in gamut
+    by construction and the recovered drives can be compared against the
+    values that produced it.
+    """
+
+    forward = emitter_matrix(primaries)
+    float_inverse = _invert_3x3(forward)
```
[the remark this answers is not in the document]
Comment (zackees) [reply ci/color_fixed_inverse_study.py]: Correct, and it was a premise error rather than a detail — thank you. Python floats are float64, so the study compared the Q16 inverse against a derivation **twice as precise** as the one it would replace.

The baseline is now modelled at the shipped precision: rounded to float32 after every operation, in the shipped operation order. Order matters as much as width here — `invert3x3` takes one reciprocal of the determinant and multiplies nine times, which is not nine divisions, and `xyY_to_XYZ` computes `1/y` once rather than dividing twice.

**It did not change the verdict, and that is worth saying plainly rather than implying the fix was decisive:**

| primaries | before (f64 baseline) | after (f32 baseline) |
|---|---|---|
| sRGB | 0 ULP | 0 ULP |
| BT.2020 | 1 ULP | 1 ULP |
| Display P3 | 1 ULP | 1 ULP |

Only the near-singular sweep moves at all — 569 ULP against 575, and 28 848 against 28 487 — and the colour error stays an order of magnitude inside A1 either way. The doc records the before/after for that reason: the answer being the same is a result, not evidence the distinction did not matter.

One thing I kept in float64 deliberately: the **rendering**. It stands in for the physical device, so rounding it to float32 would charge the emulator's own error to one of the two candidates rather than to the measurement.
